lib/Optimizer/GraphOptimizer.cpp

```diff
@@ -1807,6 +1807,37 @@ static void optimizeSliceOfSplat(Function *F) {
   }
 }
 
+/// Optimize TransposeNode into ReshapeNode when it actually moves no data.
+static void optimizeTransposeIntoReshape(Function *F) {
+  for (auto &node : F->getNodes()) {
+    auto *TR = dyn_cast<TransposeNode>(&node);
+    if (!TR)
+      continue;
+    auto inputNode = TR->getInput();
+    auto inputDims = inputNode.dims();
+    auto outputDims = TR->getResult().dims();
+    // Transpose moves no data if input/output dimensions match after they both
+    // drop dimensions of size 1. E.g. transposing [1 5 1 15] into [5 15 1 1]
```

Comment:
you do not really need additional memory here.

you could just do a pass over input and shuffle in a manner similar to merge sort.
you skip elem equals to 1 and then check if non 1 are equal from input and shuffle.

Reply:
That was the original idea but this looked like a cleaner implementation (at least code wise). I could certainly finalize that other approach if you see this as having unacceptable performance impact.

Reply:
Engineering excellence took over me. I do not think there is a significant perf impact here, considering that there are bunch of O(n*n) optimizations.

Either way is acceptable in this case, it's just I'm not sure if the other implementation is hard to follow.